grayscale.py

```diff
@@ -1,764 +1,549 @@
+                    elif (state[_ST_CONTRAST] != 0xff):
+                        contrast = state[_ST_CONTRAST]
+                        state[_ST_CONTRAST] = 0xff
+                        # Shift the value to provide 3 different levels
+                        postFrameAdj[0][1] = contrast >> 5
+                        postFrameAdj[1][1] = contrast >> 1
+                        postFrameAdj[2][1] = (contrast << 1) + 1
+                    # Check if there are pending display controller commands
+                    elif state[_ST_PENDING_CMD]:
+                        spi_write(pendingCmds)
+                        state[_ST_PENDING_CMD] = 0
+
+                # Precise idle wait for next subframe timing.
+                # This is a two stage wait for the frame time to complete.
+                # We use sleep_ms first to allow idle loop usage,
+                # with >>10 for a fast /1000 approximation.
+                # Then we use sleep_us to busy wait for the precise time.
+                sleep_ms((_FRAME_TIME - int(ticks_diff(ticks_us(), t0))) >> 10)
+                sleep_us(_FRAME_TIME - int(ticks_diff(ticks_us(), t0)))
+
+        # Announce the thread is done
+        state[_ST_THREAD] = _THREAD_STOPPED
+
+    def stopGPU(self, modeGPU=0):
+        ### Disable grayscale, stopping the running thread.
+        # If modeGPU is set to 1, it will not reset the display
+        # controller configuration.
+        ###
+        if self._state[_ST_THREAD] == _THREAD_RUNNING:
+            self._state[_ST_THREAD] = _THREAD_STOPPING
+            while self._state[_ST_THREAD] != _THREAD_STOPPED:
+                idle()
+        # Refresh the image to the B/W form
+        if not modeGPU:
+            self._displayBW.init_display()
+        self.show()
+
+
+    ## GraphicsClass functions ##
 
 
     @micropython.native
     def setFPS(self, newFrameRate):
         self.frameRate = newFrameRate
-        if newFrameRate != 0:
-            self.frameTimeMs = 1000 // newFrameRate
+
+    @micropython.viper
+    def fill(self, colour:int):
+        dBuf = ptr32(self._drawBuffer)
+        f = -1 if colour & 1 else 0
+        for i in range(_BUFF_INT_SIZE):
+            dBuf[i] = f # Black/White layer
+        f = -1 if colour & 2 else 0
+        for i in range(_BUFF_INT_SIZE, _BUFF_INT_SIZE*2):
+            dBuf[i] = f # Shading layer
+
+    def brightness(self, c):
+        if c < 0:
+            c = 0
+        elif c > 127:
+            c = 127
+        if state[_ST_THREAD] == _THREAD_RUNNING:
+            self._state[_ST_CONTRAST] = c
+        else:
+            self._displayBW.contrast(c)
 
     @micropython.native
     def update(self):
         self.show()
         if self.frameRate > 0:
-            frameTimeMs = self.frameTimeMs
-            lastUpdateEnd = self.lastUpdateEnd
-            frameTimeRemaining = frameTimeMs - utime.ticks_diff(utime.ticks_ms(), lastUpdateEnd)
+            frameTimeMs = 1000 // self.frameRate
```

Comment:
To match compatibility with the ThumbyAPI (https://github.com/TinyCircuits/TinyCircuits-Thumby-Code-Editor/blob/master/ThumbyGames/lib/thumbyGraphics.py#L55)
Note that frameRate is public and someone may be changing that in a game without calling setFPS.

Reply:
Fair enough. I remember seeing that it was done here in thumbyGraphics.py and thinking that it would be better to move it to `setFPS()`, but I appreciate the desire for parity with the ThumbyAPI.

While someone could change `frameRate` directly (rather than using `setFPS()`), it isn't clear whether that's meant to be supported or not, as the classes in the ThumbyAPI don't seem to use an underscore prefix for any properties (which I think suggests that public/private isn't being specified, not that all properties are intended to be public). But either way your change _is_ safer.